### PR TITLE
Issue #3278 Empty ResourceCollection throws NPE

### PR DESCRIPTION
**Pull request by janbartel**
Issue #3278 

When ResourceCollection has zero sized array of Resource, all of the getters throw NPE, which no callers are ever expecting to handle. Handle  zero array size same as null array, ie throw IllegalStateException.

### CODE REVIEW

**Comment by joakime** [comment]
There were a few ISE paths missing, a host of IAE paths missing, and about 8 more NPE's present in the ResourceCollection codebase.
All discovered with new Unit tests.

I've pushed a commit addressing them in this PR.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java [hunk 1]
```diff
@@ -98,17 +97,34 @@ public ResourceCollection(Resource... resources)
...
+
+        ArrayList<Resource> res = new ArrayList<>();
+
         try
         {
-            for(int i=0; i<resources.length; i++)
+            for(String strResource: resources)
+            {
+                if(strResource == null || strResource.length() == 0)
+                    continue; // skip null and empty string only (whitespace only is valid)
```
Comment (janbartel) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: I think if we're going to start testing the resource array elements, then we shouldn't accept null entries, we should throw IAE. Probably same for empty and all whitespace entries.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java [hunk 2]
```diff
@@ -148,7 +164,31 @@ public ResourceCollection(String csvResources)
...
+            _resources = null;
+            return;
+        }
+
+        List<Resource> res = new ArrayList<>();
+        for (Resource resource : resources)
+        {
+            if (resource == null)
+            {
+                continue;
```
Comment (janbartel) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: See previous comment.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java [hunk 3]
```diff
@@ -169,25 +212,38 @@ public void setResourcesAsCSV(String csvResources)
...
         {            
             while(tokenizer.hasMoreTokens())
             {
-                Resource resource = Resource.newResource(tokenizer.nextToken().trim());
-                if(!resource.exists() || !resource.isDirectory())
-                    LOG.warn(" !exist "+resource);
-                else
-                    resources.add(resource);
+                String token = tokenizer.nextToken().trim();
+                // TODO: we should not trim here, as whitespace is valid for paths
```
Comment (janbartel) [inline jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: Is a resource that is nothing but whitespace actually valid????

File: jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceCollectionTest.java [hunk 4]
```diff
@@ -18,18 +18,156 @@
...
+        assertThrows(IllegalStateException.class, ()->coll.isDirectory());
+        assertThrows(IllegalStateException.class, ()->coll.lastModified());
+        assertThrows(IllegalStateException.class, ()->coll.list());
+        assertThrows(IllegalStateException.class, ()->coll.close());
+        assertThrows(IllegalStateException.class, ()->
+        {
+            Path destPath = workdir.getPathFile("bar");
+            coll.copyTo(destPath.toFile());
+        });
+    }
```
Comment (janbartel) [inline jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceCollectionTest.java]: Good idea to have more tests.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java (hunk 1 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: Whitespace only paths are a real thing.  So are paths that start/end with whitespace (so no trimming)

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java (hunk 3 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: Yup.

Shouldn't trim for paths like 'foo ' or ' bar' too.

File: jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java (hunk 1 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: We were very inconsistent.  

For example, CSV wouldn't ISE/IAE (just log a warn)

Comment (janbartel) [reply jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java]: But this isn't a path like in the call to addPath(String path), this is adding a Resource that should have a location on disk. I can't see how a null location is wrong but an empty or whitespaceonly location is ok?

**Pull request merged**